narwhal/executor/src/notifier.rs

```diff
@@ -45,31 +44,29 @@ impl<State: ExecutionState + Send + Sync + 'static> Notifier<State> {
                 batch.transactions.len()
             );
             self.metrics.notifier_processed_batches.inc();
+
             let mut bytes = 0usize;
             for (transaction_index, transaction) in batch.transactions.into_iter().enumerate() {
                 let execution_indices = ExecutionIndices {
                     next_certificate_index: index.next_certificate_index,
                     next_batch_index: index.batch_index + 1,
                     next_transaction_index: transaction_index as u64 + 1,
+                    last_committed_leader: index.sub_dag.leader.round(),
                 };
                 bytes += transaction.len();
+
+                // NOTE: We now have access to the entire commit through `index.sub_dag`. We may
+                // thus choose to make it available to Sui through `handle_consensus_transaction`.
+
                 self.callback
-                    .handle_consensus_transaction(
-                        &index.consensus_output,
-                        execution_indices,
-                        transaction,
-                    )
+                    .handle_consensus_transaction(&index.output, execution_indices, transaction)
                     .await;
             }
-            // this is temporary
-            // we will get explicit signal from consensus on where is commit boundary in the future
-            if index.batch_index + 1
-                == index.consensus_output.certificate.header.payload.len() as u64
-            {
-                self.callback
-                    .notify_commit_boundary(&index.consensus_output)
-                    .await;
+
+            if index.sub_dag.is_last(&index.output) {
+                self.callback.notify_commit_boundary(&index.output).await;
```

Comment:
As per above, why do we feed things one by one, and then require this extra thing here to denote the last. Presumably the other side now will be re-constructing the vector, and waiting for this call for processing. Lets just send the whole commit and avoid the complexity?

Reply:
yes, I wanted to avoid touching too much at the SUI codebase in a single PR

Reply:
I should have made that clear in the PR description (I now realise)

Reply:
I also assume we are ok with the "at least once" semantics of calling the `notify_commit_boundary` method right? I mean during crash recovery this will be called again in case of commit replays

Reply:
Also looking on the increased responsibility of the `notifier` I would suggest adding some test coverage (even just the happy path)